src/Avalonia.Visuals/Matrix.cs

```diff
@@ -319,5 +319,70 @@ public static Matrix Parse(string s)
                 );
             }
         }
+
+        public static bool TryDecomposeTransform(Matrix matrix, out Decomposed decomposed)
+        {
+            decomposed = default;
+
+            var determinant = matrix.GetDeterminant();
+
+            if (determinant == 0)
+            {
+                return false;
+            }
+
+            var m11 = matrix.M11;
+            var m21 = matrix.M21;
+            var m12 = matrix.M12;
+            var m22 = matrix.M22;
+
+            // Translation.
+            decomposed.Translate = new Vector(matrix.M31, matrix.M32);
+
+            // Scale sign.
+            var scaleX = 1d;
+            var scaleY = 1d;
```

Comment:
Hello!

It would be nice to transfer the initialization to a separate `method(Matrix matrix)` and return `new Vector(scaleX, scaleY)`. There we can try to do something like this:
```
if (determinant < 0)
    if (m11 < m22)
         return new Vector(
              -Math.Sqrt(m11 * m11 + m12 * m12),
               Math.Sqrt(m21 * m21 + m22 * m22));
```

Reply:
Not sure if extracting to a separate method would be that helpful, this whole decompose method is basically `unmatrix` algorithm (https://www.w3.org/TR/css-transforms-1/#decomposing-a-2d-matrix) so perhaps it is better to keep it together. 

About adding tolerance when checking for a singular matrix - I thought about it and since both our `Matrix.HasInverse` and Chromium use exactly 0 I left it as is. I know that .net numerics package checks with epsilon so I wonder if we should add it as well.

Reply:
> this whole decompose method is basically unmatrix algorithm (https://www.w3.org/TR/css-transforms-1/#decomposing-a-2d-matrix) so perhaps it is better to keep it together.

Maybe then it's better to create an helper class? I just suggested this change to improve code readability

> About adding tolerance when checking for a singular matrix - I thought about it and since both our Matrix.HasInverse and Chromium use exactly 0 I left it as is. I know that .net numerics package checks with epsilon so I wonder if we should add it as well.

It seems to me that it is better to add, otherwise the `determinant == 0` check is pointless, as it will never be executed

Reply:
So I've added an epsilon check for inverting and decomposing. I've checked numerics package and it seems they are using different epsilon for decomposing so I used that instead.

I've tried using your helper but for me it made it less clear what this function is doing and also caused 10% perf drop unless I enabled aggressive inlining for that helper method. Could still consider moving this out, but numerics package also keeps it on `Matrix4x4` struct so I am not sure if it is worth it (@grokys - any suggestion) ?

Reply:
> So I've added an epsilon check for inverting and decomposing. I've checked numerics package and it seems they are using different epsilon for decomposing so I used that instead.

You decided to add for single. I think this is the right decision 👍

Thank you for listening!